### ISSUE TITLE
API Key auth

### ISSUE THREAD

**Issue by zhaohuabing** [envoyproxy/envoy#34877]
*Title*: *API Key auth*

*Description*:
An HTTP filter to support API Key auth.  The API key authentication filter can be used to authorize HTTP requests using the provided API key in the request. API keys are a simple and widely used method in web services to ensure that only authorized applications and users can interact with the API. It would be beneficial if Envoy could support this feature. 

An API key is a token that a client provides when making API calls. The key can be sent in the query string:

```
GET /something?api_key=abcdef12345
```

or as a request header:

```
GET /something HTTP/1.1
X-API-Key: abcdef12345
```

or as a [cookie](https://swagger.io/docs/specification/authentication/cookie-authentication/):

```
GET /something HTTP/1.1
Cookie: X-API-KEY=abcdef12345
```

[optional *Relevant Links*:]

https://swagger.io/docs/specification/authentication/api-keys/
https://github.com/envoyproxy/gateway/issues/2630

**Comment by mattklein123** [comment]
What exactly do you want this filter to do?

**Comment by zhaohuabing** [comment]
> s a token that a client provides when making API calls. The key can be sent in the query string:

Hi @mattklein123 Thanks for looking into this issue and sorry for my late response. 

The API key authentication filter can be used to authorize HTTP requests using the provided API key in the request. API keys are a simple and widely used method in web services to ensure that only authorized applications and users can interact with the API. It would be beneficial if Envoy could support this feature.

**Comment by sanposhiho** [comment]
Hi 👋 
I'm coming from https://github.com/envoyproxy/gateway/issues/2630, interested in realizing it.
Actually I'm researching other competing gateways, and actually Key auth functionality is one of the gaps that some other implementations implement, while the envoy doesn't. So, I believe it's worthy enough.

I can take an implementer part, would any maintainer be willing to help in reviewing etc to promote it?

**Comment by gecube** [comment]
It is very pity that we don’t have reliable and out-of-box way to use api keys in envoy proxy.

**Comment by arkodg** [comment]
although API Key Auth is not the most secure or the most performant way to do authn & authz, its still used in existing brownfield environments and also as an additional auth mechanism like the one outlined [here](https://cloud.google.com/endpoints/docs/openapi/restricting-api-access-with-api-keys), so having support for it would unblock migration to envoy proxy.
Here are the requirements that necessitates a new filter instead of reusing an existing one like [the RBAC filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rbac_filter)
* Ability to read the API key from a [header, cookie, or query param](https://swagger.io/docs/specification/v3_0/authentication/api-keys/)
* Provide a list/map or allowed API Keys similar to [what BasicAuth has](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/basic_auth/v3/basic_auth.proto#extensions-filters-http-basic-auth-v3-basicauth)
* Provide a way to forward the Client id linked to the the API key similar to [what Basic Auth has](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/basic_auth/v3/basic_auth.proto#extensions-filters-http-basic-auth-v3-basicauth)

**Comment by gecube** [comment]
Agree, that it is essential functionality.

**Comment by wbpcode** [comment]
I agree this is essential functionality. I can help with the review if someone want to contribute.

**Comment by sanposhiho** [comment]
Thanks @wbpcode! I'll ping you when I create the PR.
